src/main/java/com/exasol/adapter/dialects/rls/RowLevelSecurityQueryRewriter.java

```diff
@@ -133,69 +108,119 @@ private SqlSelectList getSqlSelectList(final SqlStatementSelect select) {
+        final SqlStatementSelect.Builder rlsStatementBuilder = SqlStatementSelect.builder().selectList(sqlSelectList)
+                .fromClause(select.getFromClause());
+        if (select.hasGroupBy()) {
+            rlsStatementBuilder.groupBy(select.getGroupBy());
+        }
+        if (select.hasHaving()) {
+            rlsStatementBuilder.having(select.getHaving());
+        }
+        if (select.hasLimit()) {
+            rlsStatementBuilder.limit(select.getLimit());
+        }
+        if (select.hasOrderBy()) {
+            rlsStatementBuilder.orderBy(select.getOrderBy());
+        }
+        return rlsStatementBuilder;
+    }
+
     private SqlNode createWhereClause(final SqlStatementSelect select, final UserInformation userInformation,
-            final boolean protectedWithExaRowRoles, final boolean protectedWithExaRowTenants) throws SQLException {
-        final Optional<SqlNode> whereClauseForRoles = setWhereClauseForRoles(protectedWithExaRowRoles, userInformation);
-        final Optional<SqlNode> whereClauseForTenants = setWhereClauseForTenants(protectedWithExaRowTenants,
-                userInformation);
-        final List<SqlNode> arguments = new ArrayList<>(3);
+            final TableProtectionDetails protection) throws SQLException {
         if (select.hasFilter()) {
-            arguments.add(select.getWhereClause());
-        }
-        whereClauseForRoles.ifPresent(arguments::add);
-        whereClauseForTenants.ifPresent(arguments::add);
-        if ((arguments.size() == 2) || (arguments.size() == 3)) {
-            return new SqlPredicateAnd(arguments);
-        } else if (arguments.size() == 1) {
-            return arguments.get(0);
+            return combineRegularFilterWithRlsFilter(select, userInformation, protection);
         } else {
-            throw new IllegalArgumentException("Unexpected size of the result set.");
+            return createRlsFilter(userInformation, protection);
         }
     }
 
-    private Optional<SqlNode> setWhereClauseForTenants(final boolean protectedWithExaRowTenants,
-            final UserInformation userInformation) {
-        if (protectedWithExaRowTenants) {
-            return Optional.of(getExaRowTenantsNode(userInformation));
+    private SqlNode combineRegularFilterWithRlsFilter(final SqlStatementSelect select,
+            final UserInformation userInformation, final TableProtectionDetails protection) throws SQLException {
+        final SqlNode where = select.getWhereClause();
+        if (protection.isTenantProtected()) {
+            if (protection.isRoleProtected()) {
+                return and(where, createTenantsNode(userInformation), createRolesNode(userInformation));
+            } else if (protection.isGroupProtected()) {
+                return and(where, or(createTenantsNode(userInformation), createGroupNode(userInformation)));
+            } else {
+                return and(where, createTenantsNode(userInformation));
```

Comment:
I was thinking about extracting a method, but then you lose the context of why combinations appear in the tenant part and not in the others (because that has already been checked before).

The main reason I didn't get it more pretty is our "favorite" SqlNode object hierarchy.